Peer data

vaeryn-uk opened the pull request:
Hi @codecat , thanks for the work with these bindings. Much appreciated.

I've made a few changes here if you are interested in bringing them upstream?
* Exposed `ENetPeer::data`, http://enet.bespin.org/structENetPeer.html#a1873959810db7ac7a02da90469ee384e
* Added a test to cover this (I'm new to cgo, so wanted to get coverage here).
* ~Exposed a function for the enet version, which may be useful for debugging.~  #7 
* ~Added an [Godoc-compliant example](https://pkg.go.dev/github.com/vaeryn-uk/go-enet#example-package-ClientServer) based on the example already in the readme for some extra test coverage.~  #7 
* ~Added a docker image & makefile to run the tests, with a [Github workflow](https://github.com/vaeryn-uk/go-enet/actions/runs/3347668767/jobs/5545907003).~  #7 
* ~Removed the enet source as I'm not sure this is actually used for anything, other than maybe a reference? As the enet version is quickly modified in the dockerfile, I'm not sure having a copy of the (potentially old) source is necessary?~
* ~`go.mod`~ #7 

Let me know if you're interested in some parts of this, but not others. Happy to remove anything that doesn't fit.

codecat commented:
The source (actually just header files and library) exists to make sure it builds on Windows (see `enetc.go` which includes it). With these changes, I'm pretty sure this will only build on Linux, Mac, and Windows **if** the `enet` folder can be found in `PATH`.

vaeryn-uk commented:
That makes sense, thanks. I've restored the enet headers & lib.

codecat commented:
Make sure you restore the `.gitattributes` file too!

vaeryn-uk commented:
Hi @codecat . Are you still interested in this PR? Let me know if you'd like any more changes. Thanks.

codecat commented on peer.go:
```diff
@@ -71,3 +95,35 @@ func (peer enetPeer) SendPacket(packet Packet, channel uint8) error {
+	b := make([]byte, len(data)+1)
+	b[0] = byte(len(data))
+	copy(b[1:], data)
+	peer.cPeer.data = unsafe.Pointer(&b[0])
```
How long does the data stay valid for? I assume there will come a point where `b` gets freed by Go. Does the data need to be stored in the `enetPeer` object for the unsafe pointer to remain valid?

I know that in the tests you check for validity after a GC call, but is that guaranteed? Could it still point to memory that is "valid data" but still freed memory?

codecat replied:
Additionally, this limits the data length to 255 bytes, no?

vaeryn-uk replied:
My understanding is that because Go created both the byte slice and unsafe pointer, GC won't touch it for as long as the pointer lives (from some reading [here](https://groups.google.com/g/golang-nuts/c/yNis7bQG_rY/m/yaJFoSx1hgIJ)). My thinking was that putting this in the C struct (`peer.cPeer.data`) would keep the pointer alive as far as Go's GC is concerned, but now I'm not sure. This is why I covered it in the test, but yes you're quite right that maybe this is passing accidentally. I'll need to do more reading to comment more confidently.

vaeryn-uk replied:
Yes, this has a 255 limit which is mentioned in the interface method docs, but as per https://github.com/codecat/go-enet/pull/5/files/24b0625eda56718d111e197e8c0e65f766df8aab#r1060198005 I'll see if I can come up with something a little less limiting.

vaeryn-uk replied:
> How long does the data stay valid for? I assume there will come a point where b gets freed by Go. Does the data need to be stored in the enetPeer object for the unsafe pointer to remain valid?

I've addressed this in my latest change and explained my thinking in [the commit message](https://github.com/vaeryn-uk/go-enet/commit/951bfb7266d6a7f1e27a98aa3a3a629149533bf4).

Reading https://pkg.go.dev/cmd/cgo, using `C.Cbytes` ensures this data will stay valid in C's heap and out of reach of Go's GC. As mentioned in the commit, I couldn't find a neat way to auto-free this, so unfortunately the best I can think of is having the user do it.

> I know that in the tests you check for validity after a GC call, but is that guaranteed? Could it still point to memory that is "valid data" but still freed memory?

You were absolutely right, it was not guaranteed which `cgocheck` highlighted.

> Additionally, this limits the data length to 255 bytes, no?

I've increased this so the limit should be sufficient for any practical use-case.

codecat commented on peer.go:
```diff
@@ -71,3 +95,35 @@ func (peer enetPeer) SendPacket(packet Packet, channel uint8) error {
+
+	return C.GoBytes(
+		// Skip the first byte as this is the slice length.
+		unsafe.Add(ptr, 1),
```
This (to me) implies we might not want to use slices for this, but I'm not too familiar with the `unsafe` package.

vaeryn-uk replied:
Hmm good point. At the time I wanted something that I could easily convert between C and go, but actually as this is all in Go memory, I'm now wondering if this couldn't just be any data that you wanted to pass (i.e. `any`). My memory is a little hazy.

I'll go and try and come up with something better and/or find better justifications for what's here.

Thanks for taking the time for a review. Appreciated.

vaeryn-uk replied:
I spent a bit of time playing around with the idea of these functions accepting/returning `any`, but couldn't find a neat way to do this without violating cgo's memory rules (i.e. don't store Go pointers in C heap). I like the slice here because of the `C.CBytes` and `C.GoBytes` helpers. It did cross my mind to try serialise `any` in to `bytes` ([gob?](https://pkg.go.dev/encoding/gob)) before putting in C, but I wasn't sure this is something I wanted to get in to for this API - feels like a bit of a rabbit hole.

My use case for these changes is to store an application identifier alongside the peer; essentially my calling code would be `ev.GetPeer().SetData([]byte(userId))`. I feel like the `[]byte` is simple & flexible.

codecat replied:
Indeed, I think `[]byte` is sufficient here. This lets the user decide by themselves how stuff is serialized (for example using json)

vaeryn-uk commented:
@codecat , I've taken another swing at this but I'd like to spend a bit more time testing this in my application before I'm happy my changes work as I intend them to. I just wanted to add replies now whilst my learnings are still fresh on my mind :)

vaeryn-uk commented:
> @codecat , I've taken another swing at this but I'd like to spend a bit more time testing this in my application before I'm happy my changes work as I intend them to. I just wanted to add replies now whilst my learnings are still fresh on my mind :)

Sorry for the delay. I've done testing and happy with the changes as are if you'd like to take another look? Thanks.

codecat commented:
Do you think you can separate this into several PR's? It looks like several things in 1:

* Adding tests, docker build, and Github actions
* Adding `LinkedVersion`
* Adding `Peer.SetData` and `Peer.GetData`

I think the tests, docker build, github actions, and maybe `LinkedVersion` can be 1 PR, but the `SetData` and `GetData` should be 1 PR as well. I think having the big feature (peer data) implemented separately from the tests is good practice for keeping proper Git history.

vaeryn-uk committed:
Expose enet peer data field

Decided to allow storing of arbitrary data saved as a byte slice,
leaving the conversion of whatever data that wants to be stored (e.g. an
identified or JSON) up to the consumer.

Importantly this data is being stored in enet C field so we ensure that
the data is being copied out in to C allocation, rather than allowing C
code to have Go pointers. For this reason, it is up to the caller to
unset data when the peer is done with to avoid memory leaks. I couldn't
find a way with enet to have this automatically cleaned up.

vaeryn-uk commented:
Okay, done. I've created the separate PR as requested: https://github.com/codecat/go-enet/pull/7

Then I'll leave this as just the implementation of the peer data. I've cleaned up the commit history too. Note this includes the CI changes too as I presume you'll want to merge that one first, then this one.

codecat commented:
Thanks! This PR still seems to contain the same changes as in #7 though.

vaeryn-uk commented:
Yup, my thinking was that you'd want to merge the other one first, and that would just leave the additional peer-data changes here to avoid a possible merge commit. But no problem; I have now removed those commits from this PR.

codecat commented:
Ahh, I see. Yeah I wanted to make sure the diff of this PR is clear and concise. 👍

codecat commented:
Thank you 😊
